adapter/memQuota/memQuota.go

```diff
@@ -0,0 +1,296 @@
+)
+
+const (
+	ticksPerSecond = 10
+)
+
+// Register records the builders exposed by this adapter.
+func Register(r adapter.Registrar) {
+	r.RegisterQuota(newBuilder())
+}
+
+func newBuilder() builder {
+	return builder{adapter.NewDefaultBuilder(name, desc, conf)}
+}
+
+func (builder) NewQuota(env adapter.Env, c adapter.AspectConfig, d map[string]*adapter.QuotaDefinition) (adapter.QuotaAspect, error) {
+	return newAspect(env, c.(*config.Params), d)
+}
+
+// newAspect returns a new aspect.
+func newAspect(env adapter.Env, c *config.Params, definitions map[string]*adapter.QuotaDefinition) (adapter.QuotaAspect, error) {
+	mq := &memQuota{
+		definitions: definitions,
+		cells:       make(map[string]int64),
+		windows:     make(map[string]*rollingWindow),
+		recentDedup: make(map[string]int64),
+		oldDedup:    make(map[string]int64),
+		ticker:      time.NewTicker(time.Duration(c.MinDeduplicationWindowSeconds) * time.Second),
+		getTick:     func() int32 { return int32(time.Now().UnixNano() / (1000000000 / ticksPerSecond)) },
+		logger:      env.Logger(),
+	}
+
+	go func() {
+		for range mq.ticker.C {
+			mq.reapDedup()
+		}
+	}()
+
+	return mq, nil
+}
+
+func (mq *memQuota) Close() error {
+	mq.ticker.Stop()
+	return nil
+}
+
+func (mq *memQuota) Alloc(args adapter.QuotaArgs) (int64, error) {
+	return mq.alloc(args, false)
+}
+
+func (mq *memQuota) AllocBestEffort(args adapter.QuotaArgs) (int64, error) {
+	return mq.alloc(args, true)
+}
+
+func (mq *memQuota) alloc(args adapter.QuotaArgs, bestEffort bool) (int64, error) {
+	return mq.common(args, func(d *adapter.QuotaDefinition, key string) int64 {
+		result := args.QuotaAmount
+
+		// we optimize storage for non-expiring quotas
+		if d.ExpirationSeconds == 0 {
+			inUse := mq.cells[key]
+
+			if result > d.MaxAmount-inUse {
+				if bestEffort {
+					// grab as much as we can
+					result = d.MaxAmount - inUse
+				} else {
+					result = 0
+				}
+			}
+
+			if result > 0 {
+				mq.cells[key] = inUse + result
+			}
+		} else {
+			window, ok := mq.windows[key]
+			if !ok {
+				window = newRollingWindow(d.MaxAmount, d.ExpirationSeconds*ticksPerSecond)
+				mq.windows[key] = window
+			}
+
+			currentTick := mq.getTick()
+			if !window.alloc(result, currentTick) {
+				if bestEffort {
+					// grab as much as we can
+					result = window.available()
+					_ = window.alloc(result, currentTick)
+				} else {
+					result = 0
+				}
+			}
+		}
+
+		return result
+	})
+}
+
+func (mq *memQuota) ReleaseBestEffort(args adapter.QuotaArgs) (int64, error) {
+	return mq.common(args, func(d *adapter.QuotaDefinition, key string) int64 {
+		result := args.QuotaAmount
+
+		if d.ExpirationSeconds == 0 {
+			inUse := mq.cells[key]
+
+			if result == inUse {
+				// delete the cell since it contains no useful state
+				delete(mq.cells, key)
+			} else {
+				if result > inUse {
+					// free as much as we can
+					result = inUse
+				}
+
+				mq.cells[key] = inUse - result
+			}
+		} else {
```

Comment:
can we rewrite this as:

```
if d.ExpirationSeconds > 0 {
   return 0
}
....
```

(dropping the `else` and pulling the no-op condition up front).